src/GuzzleRequestAdapter.php

```diff
@@ -281,152 +356,282 @@ public function getBaseUrl(): string
+        $scope = $serializationSpan->activate();
+        try {
+            if (!$response->hasHeader(RequestInformation::$contentTypeHeader)) {
+                throw new RuntimeException("No response content type header for deserialization");
+            }
+            $contentType = explode(';', $response->getHeaderLine(RequestInformation::$contentTypeHeader));
+
+            $result = $this->parseNodeFactory->getRootParseNode($contentType[0], $response->getBody());
+            $serializationSpan->setStatus(StatusCode::STATUS_OK, 'deserialize_success');
+            return $result;
+        } finally {
+            $scope->detach();
+            $serializationSpan->end();
         }
-        $contentType = explode(';', $response->getHeaderLine(RequestInformation::$contentTypeHeader));
-        return $this->parseNodeFactory->getRootParseNode($contentType[0], $response->getBody());
     }
 
     /**
      * Authenticates and executes the request
      *
      * @param RequestInformation $requestInformation
      * @param string $claims additional claims to request if CAE fails
+     * @param SpanInterface|null $span
      * @return Promise
      */
-    private function getHttpResponseMessage(RequestInformation $requestInformation, string $claims = ""): Promise
+    private function getHttpResponseMessage(RequestInformation $requestInformation, string $claims = '', ?SpanInterface $span = null): Promise
     {
-        $requestInformation->pathParameters['baseurl'] = $this->getBaseUrl();
-        $additionalAuthContext = $claims ? ['claims' => $claims] : [];
-        $request = $this->authenticationProvider->authenticateRequest($requestInformation, $additionalAuthContext);
-        return $request->then(
-            function () use ($requestInformation) {
-                $psrRequest = $this->getPsrRequestFromRequestInformation($requestInformation);
-                return $this->guzzleClient->send($psrRequest, $requestInformation->getRequestOptions());
-            }
-        )->then(
-            function (ResponseInterface $response) use ($requestInformation, $claims) {
-                return $this->retryCAEResponseIfRequired($response, $requestInformation, $claims);
-            }
-        );
+        $httpResponseSpan = $this->tracer->spanBuilder('getHttpResponseMessage');
+        if ($span !== null) {
+            $httpResponseSpan = $httpResponseSpan->addLink($span->getContext())->setParent($span->storeInContext(Context::getCurrent()));
+        }
+        $httpResponseSpan = $httpResponseSpan->startSpan();
+        $scope = $httpResponseSpan->activate();
+        try {
+            $requestInformation->pathParameters['baseurl'] = $this->getBaseUrl();
+            $additionalAuthContext                         = $claims ? ['claims' => $claims] : [];
+            $request                                       = $this->authenticationProvider->authenticateRequest($requestInformation, $additionalAuthContext);
```

Comment:
```suggestion
            $request = $this->authenticationProvider->authenticateRequest($requestInformation, $additionalAuthContext);
```
Nit; affects readability IMO